File: GeoBroker-Client/src/main/java/de/hasenburg/geobroker/client/main/SimpleClient.java
package de.hasenburg.geobroker.client.main;

import de.hasenburg.geobroker.commons.Utility;
import de.hasenburg.geobroker.commons.communication.ZMQControlUtility;
import de.hasenburg.geobroker.commons.communication.ZMQProcessManager;
import de.hasenburg.geobroker.commons.model.KryoSerializer;
import de.hasenburg.geobroker.commons.model.message.ControlPacketType;
import de.hasenburg.geobroker.commons.model.message.ReasonCode;
import de.hasenburg.geobroker.client.communication.ZMQProcessStarter;
import de.hasenburg.geobroker.client.communication.ZMQProcess_SimpleClient;
import de.hasenburg.geobroker.client.communication.InternalClientMessage;
import de.hasenburg.geobroker.commons.model.message.payloads.CONNECTPayload;
import de.hasenburg.geobroker.commons.model.message.payloads.DISCONNECTPayload;
import de.hasenburg.geobroker.commons.model.spatial.Location;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.Nullable;
import org.zeromq.SocketType;
import org.zeromq.ZMQ;
import org.zeromq.ZMsg;

import java.util.Optional;
import java.util.Random;

public class SimpleClient {

	private static final Logger logger = LogManager.getLogger();

	private ZMQProcessManager processManager;
	private String identifier;
	private KryoSerializer kryo = new KryoSerializer();
	ZMQ.Socket orderSocket;


	public SimpleClient(@Nullable String identifier, String address, int port, ZMQProcessManager processManager) {
		if (identifier == null) {
			Random random = new Random();
			identifier = "SimpleClient-" + System.nanoTime();
		}

		this.identifier = identifier;
		this.processManager = processManager;
		ZMQProcessStarter.runZMQProcess_SimpleClient(processManager, address, port, identifier);
		orderSocket = processManager.getContext().createSocket(SocketType.REQ);
		orderSocket.setIdentity(identifier.getBytes());
		orderSocket.connect(Utility.generateClientOrderBackendString(identifier));

		logger.info("Created client {}", identifier);
	}

	public String getIdentity() {
		return identifier;
	}

	public void tearDownClient() {
		orderSocket.setLinger(0);
		processManager.getContext().destroySocket(orderSocket);
		processManager.sendCommandToZMQProcess(getIdentity(), ZMQControlUtility.ZMQControlCommand.KILL);
	}

	public ZMsg sendInternalClientMessage(InternalClientMessage message) {
		ZMsg orderMessage = ZMsg.newStringMsg(ZMQProcess_SimpleClient.ORDERS.SEND.name());
		ZMsg internalClientMessage = message.getZMsg(kryo);
		for (int i = 0; i <= internalClientMessage.size(); i++) {
			orderMessage.add(internalClientMessage.pop());
		}

		orderMessage.send(orderSocket);
		return (ZMsg.recvMsg(orderSocket));
	}

	public InternalClientMessage receiveInternalClientMessage() {
		ZMsg orderMessage = ZMsg.newStringMsg(ZMQProcess_SimpleClient.ORDERS.RECEIVE.name());

		// send order
		orderMessage.send(orderSocket);
		final Optional<InternalClientMessage> clientMessageO = InternalClientMessage.buildMessage(ZMsg.recvMsg(
				orderSocket), kryo);

		return clientMessageO.orElse(null);
	}

	/**
	 * @param - receive timeout in ms
	 * @return a message from the server or null if server did not send a message
	 */
	public @Nullable InternalClientMessage receiveInternalClientMessageWithTimeout(int timeout) {
		ZMsg orderMessage = ZMsg.newStringMsg(ZMQProcess_SimpleClient.ORDERS.RECEIVE_WITH_TIMEOUT.name(), timeout + "");

		// send order
		orderMessage.send(orderSocket);
		ZMsg response = ZMsg.recvMsg(orderSocket);
		// check first frame as might be empty
		if (ZMQProcess_SimpleClient.ORDERS.EMPTY.name().equals(response.getFirst().getString(ZMQ.CHARSET))) {
			return null;
		}

		final Optional<InternalClientMessage> clientMessageO = InternalClientMessage.buildMessage(response, kryo);

		return clientMessageO.orElse(null);
	}

	public static void main(String[] args) {
		ZMQProcessManager processManager = new ZMQProcessManager();
		SimpleClient client = new SimpleClient(null, "localhost", 5559, processManager);

		// connect
		InternalClientMessage clientMessage = new InternalClientMessage(ControlPacketType.CONNECT,
				new CONNECTPayload(Location.random()));
		client.sendInternalClientMessage(clientMessage);

		// receive one message
		InternalClientMessage response = client.receiveInternalClientMessage();
		logger.info("Received server answer: {}", response);

		// wait 5 seconds
		Utility.sleepNoLog(5000, 0);

		// disconnect
		clientMessage = new InternalClientMessage(ControlPacketType.DISCONNECT,
				new DISCONNECTPayload(ReasonCode.NormalDisconnection));
		client.sendInternalClientMessage(clientMessage);

		client.tearDownClient();
		if (processManager.tearDown(3000)) {
			logger.info("SimpleClient shut down properly.");
		} else {
			logger.fatal("ProcessManager reported that processes are still running: {}",
					processManager.getIncompleteZMQProcesses());
		}
		System.exit(0);
	}


}

File: GeoBroker-Client/src/main/kotlin/de/hasenburg/geobroker/client/main/SimpleClient.kt
package de.hasenburg.geobroker.client.main

import de.hasenburg.geobroker.commons.communication.SPDealer
import de.hasenburg.geobroker.commons.communication.ZMQProcessManager
import de.hasenburg.geobroker.commons.model.message.*
import de.hasenburg.geobroker.commons.model.message.Payload.*
import de.hasenburg.geobroker.commons.model.spatial.Geofence
import de.hasenburg.geobroker.commons.model.spatial.Location
import de.hasenburg.geobroker.commons.sleepNoLog
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withTimeoutOrNull
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonConfiguration
import org.apache.logging.log4j.LogManager
import kotlin.system.exitProcess

private val logger = LogManager.getLogger()

/**
 * The SimpleClient connects to the GeoBroker running at [ip]:[port].
 * If an [identity] is provided on startup, it announces itself with this identify; otherwise, a default identify is used.
 *
 * It is possible to supply a [socketHWM], for more information on HWM, check out the ZeroMQ documentation.
 * If no HWM is supplied, 1000 is used.
 */
class SimpleClient(ip: String, port: Int, socketHWM: Int = 1000, val identity: String = "SimpleClient-" + System.nanoTime()) {

    private val spDealer = SPDealer(ip, port, socketHWM)
    private val json = Json(JsonConfiguration.Stable)

    fun tearDownClient() {
        if (spDealer.isActive) {
            spDealer.shutdown()
        }
    }

    /**
     * Send the given [payload] to the broker.
     * Returns true, if successful, otherwise false.
     */
    fun send(payload: Payload): Boolean {
        val zMsg = payload.toZMsg(json, clientIdentifier = identity)
        return spDealer.toSent.offer(zMsg)
    }

    /**
     * Receives a message from the broker, blocks until a message was received.
     * Then, it returns the [Payload] of the message.
     */
    fun receive(): Payload {
        return runBlocking {
            val zMsgTP = spDealer.wasReceived.receive()
            zMsgTP.msg.toPayloadAndId(json)!!.second
        }
    }

    /**
     * Receives a message from the blocker, blocks as defined by [timeout] in ms.
     * Then, it returns the [Payload] or the message or null, if none was received.
     **/
    fun receiveWithTimeout(timeout: Int): Payload? {
        return runBlocking {
            withTimeoutOrNull(timeout.toLong()) {
                val zMsgTP = spDealer.wasReceived.receive()
                zMsgTP.msg.toPayloadAndId(json)!!.second
            }
        }
    }

}

fun main() {
    val processManager = ZMQProcessManager()
    val client = SimpleClient("localhost", 5559)

    // connect
    client.send(CONNECTPayload(Location.random()))

    // receive one message
    logger.info("Received server answer: {}", client.receive())

    // subscribe
    client.send(SUBSCRIBEPayload(Topic("test"), Geofence.circle(Location.random(), 2.0)))

    // receive one message
    logger.info("Received server answer: {}", client.receive())

    // wait 5 seconds
    sleepNoLog(5000, 0)

    // disconnect
    client.send(DISCONNECTPayload(ReasonCode.NormalDisconnection))

    client.tearDownClient()
    if (processManager.tearDown(3000)) {
        logger.info("SimpleClient shut down properly.")
    } else {
        logger.fatal("ProcessManager reported that processes are still running: {}",
                processManager.incompleteZMQProcesses)
    }
    exitProcess(0)
}